{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Rhyzopertha dominica"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## get data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "data were shared by Oppert, Brenda <Brenda.Oppert@ARS.USDA.GOV>\n",
    "\n",
    "download the genome and RNA-seq data from https://de.cyverse.org/de/ with Cyberduck. The software Cyberduck is not user-friendly. The downloaded files are easily get broken, many times larger than expected. Too much errors. Need to be careful with it."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### fix sequence problem\n",
    "get a error while running Trinity, in the Trimmamatic step, base length and quality length do not match for file `Adult_filtered_6-25-2012-USDA_NoIndex_L005_R1.fastq` and `Adult_filtered_6-25-2012-USDA_NoIndex_L005_R2.fastq`. Fix that.\n",
    "\n",
    "need to be very careful when downloading the file with the Cyberduck program.\n",
    "files\n",
    "```\n",
    "Adult_filtered_6-25-2012-USDA_NoIndex_L005_R1.fastq\n",
    "Adult_filtered_6-25-2012-USDA_NoIndex_L005_R2.fastq\n",
    "Carcass-1_S3_L001_R1_001.fastq\n",
    "Carcass-1_S3_L001_R2_001.fastq\n",
    "Carcass-2_S5_L001_R1_001.fastq\n",
    "Carcass-2_S5_L001_R2_001.fastq\n",
    "Carcass-3_S9_L001_R1_001.fastq\n",
    "Carcass-3_S9_L001_R2_001.fastq\n",
    "Gut-1_S2_L001_R1_001.fastq\n",
    "Gut-1_S2_L001_R2_001.fastq\n",
    "Gut-2_S6_L001_R1_001.fastq\n",
    "Gut-2_S6_L001_R2_001.fastq\n",
    "Gut-3_S8_L001_R1_001.fastq\n",
    "Gut-3_S8_L001_R2_001.fastq\n",
    "Head-1_S1_L001_R1_001.fastq\n",
    "Head-1_S1_L001_R2_001.fastq\n",
    "Head-2_S4_L001_R1_001.fastq\n",
    "Head-2_S4_L001_R2_001.fastq\n",
    "Head-3_S7_L001_R1_001.fastq\n",
    "Head-3_S7_L001_R2_001.fastq\n",
    "ResExp_Rep2.fastq\n",
    "ResExp_Rep3a.fastq\n",
    "ResExp_Rep3b.fastq\n",
    "ResExp_Rep3c.fastq\n",
    "ResExp_Rep4.fastq\n",
    "ResUxp_Rep2.fastq\n",
    "ResUxp_Rep3.fastq\n",
    "ResUxp_Rep4.fastq\n",
    "SRR1130516_1.fastq\n",
    "SRR1130516_2.fastq\n",
    "SuscExp_Rep2.fastq\n",
    "SuscExp_Rep3.fastq\n",
    "SuscExp_Rep4.fastq\n",
    "SuscUxp_Rep2.fastq\n",
    "SuscUxp_Rep3.fastq\n",
    "SuscUxp_Rep4.fastq\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## de novo assembly"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Trinity"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "transfer the genome and RNA-seq data to Elf, folder `/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "srun -N 1 --partition long --qos long-award   -t 240:00:00 --pty bash #allocate one node for running Trinity\n",
    "cd /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA\n",
    "/gpfs/gpfs/scratch/xc278/p/trinityrnaseq-Trinity-v2.8.5/Trinity --seqType fq --max_memory 200G --single Adult_filtered_6-25-2012-USDA_NoIndex_L005_R1.fastq,Carcass-1_S3_L001_R1_001.fastq,Carcass-2_S5_L001_R1_001.fastq,Carcass-3_S9_L001_R1_001.fastq,Gut-1_S2_L001_R1_001.fastq,Gut-2_S6_L001_R1_001.fastq,Gut-3_S8_L001_R1_001.fastq,Head-1_S1_L001_R1_001.fastq,Head-2_S4_L001_R1_001.fastq,Head-3_S7_L001_R1_001.fastq,ResExp_Rep2.fastq,ResExp_Rep3a.fastq,ResExp_Rep3b.fastq,ResExp_Rep3c.fastq,ResExp_Rep4.fastq,ResUxp_Rep2.fastq,ResUxp_Rep3.fastq,ResUxp_Rep4.fastq,SuscExp_Rep2.fastq,SuscExp_Rep3.fastq,SuscExp_Rep4.fastq,SuscUxp_Rep2.fastq,SuscUxp_Rep3.fastq,SuscUxp_Rep4.fastq,Adult_filtered_6-25-2012-USDA_NoIndex_L005_R2.fastq,Carcass-1_S3_L001_R2_001.fastq,Carcass-2_S5_L001_R2_001.fastq,Carcass-3_S9_L001_R2_001.fastq,Gut-1_S2_L001_R2_001.fastq,Gut-2_S6_L001_R2_001.fastq,Gut-3_S8_L001_R2_001.fastq,Head-1_S1_L001_R2_001.fastq,Head-2_S4_L001_R2_001.fastq,Head-3_S7_L001_R2_001.fastq --CPU 36 --trimmomatic --no_normalize_reads --output /home/scratch/Trinity/ --min_kmer_cov 2 --run_as_paired\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "when running Trinity, the sequences names are not accepted by Trinity. Change with the scripts like\n",
    "```\n",
    "zcat input_reverse.fastq.gz | awk '{{print (NR%4 == 1) ? \"@1_\" ++i \"/2\": $0}}' | gzip -c > output_reverse.renamed.fastq.gz\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### binpacker"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "copy the combined single.fa generated by Trinity to folder /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA\n",
    "\n",
    "```bash\n",
    "srun -N 1 --partition long --qos long-award   -t 240:00:00 --pty bash #allocate one node for running BinPacker\n",
    "cd /home/scratch\n",
    "cp /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/single.fa.gz ./\n",
    "pigz -d single.fa.gz\n",
    "/gpfs/gpfs/scratch/xc278/p/BinPacker/BinPacker_binary/BinPacker -s fa -p single -o ./BinPacker/ -u single.fa\n",
    "\n",
    "\n",
    "```\n",
    "\n",
    "conclusion: do not use BinPacker. It does not support multiple threads and would be slow. It is not updated for a long time."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### rnaSPAdes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```bash\n",
    "srun -N 1 --partition long --qos long-award   -t 240:00:00 --pty bash #allocate one node for running rnaSPAdes\n",
    "cd /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA\n",
    "/home1/xc278/w/scratch/p/SPAdes/SPAdes-3.12.0-Linux/bin/rnaspades.py -o /home/scratch/SPAdes/ -t 36 -m 2000 --pe1-1 Adult_filtered_6-25-2012-USDA_NoIndex_L005_R1.fastq --pe1-2 Adult_filtered_6-25-2012-USDA_NoIndex_L005_R2.fastq --pe2-1 Carcass-1_S3_L001_R1_001.fastq --pe2-2 Carcass-1_S3_L001_R2_001.fastq --pe2-1 Carcass-2_S5_L001_R1_001.fastq --pe2-2 Carcass-2_S5_L001_R2_001.fastq --pe2-1 Carcass-3_S9_L001_R1_001.fastq --pe2-2 Carcass-3_S9_L001_R2_001.fastq --pe3-1 Gut-1_S2_L001_R1_001.fastq --pe3-2 Gut-1_S2_L001_R2_001.fastq --pe3-1 Gut-2_S6_L001_R1_001.fastq --pe3-2 Gut-2_S6_L001_R2_001.fastq --pe3-1 Gut-3_S8_L001_R1_001.fastq --pe3-2 Gut-3_S8_L001_R2_001.fastq --pe4-1 Head-1_S1_L001_R1_001.fastq --pe4-2 Head-1_S1_L001_R2_001.fastq --pe4-1 Head-2_S4_L001_R1_001.fastq --pe4-2 Head-2_S4_L001_R2_001.fastq --pe4-1 Head-3_S7_L001_R1_001.fastq --pe4-2 Head-3_S7_L001_R2_001.fastq --s1 ResExp_Rep2.fastq --s1 ResExp_Rep3a.fastq --s1 ResExp_Rep3b.fastq --s1 ResExp_Rep3c.fastq --s1 ResExp_Rep4.fastq --s1 ResUxp_Rep2.fastq --s1 ResUxp_Rep3.fastq --s1 ResUxp_Rep4.fastq --s1 SuscExp_Rep2.fastq --s1 SuscExp_Rep3.fastq --s1 SuscExp_Rep4.fastq --s1 SuscUxp_Rep2.fastq --s1 SuscUxp_Rep3.fastq --s1 SuscUxp_Rep4.fastq\n",
    "\n",
    "```\n",
    "\n",
    "do not use rnaSPAdes, too slow."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Bridger\n",
    "```\n",
    "/gpfs/gpfs/scratch/xc278/p/bridger/Bridger_r2014-12-01/Bridger.pl --seqType fa --single /home/scratch/single.fa.gz -o /home/scratch/Bridger --CPU 36\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### transAbyss"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/home1/xc278/p/anaconda3/lib/\n",
    "/home1/xc278/p/anaconda3/envs/bio/bin/transabyss --se single.fa --length 200 --threads 36 --outdir transabyss --mpi 36 --pe /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/SRR1130516_1.fastq /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/SRR1130516_2.fastq\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### conclusion for de novo\n",
    "\n",
    "Trinity and TransAbyss are faster. Use them."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## HISTAT2 StringTie"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```bash\n",
    "cd /home/scratch\n",
    "#build genome\n",
    "/gpfs/gpfs/scratch/xc278/p/HISTAT2/hisat2-2.1.0/hisat2-build -p 36 /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/Genome/Rdo_Scaffolds.fa Rdo #build genome, 36 core\n",
    "#align with HISTAT2\n",
    "/gpfs/gpfs/scratch/xc278/p/HISTAT2/hisat2-2.1.0/hisat2 -p 36 -x Rdo -1 /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/Adult_filtered_6-25-2012-USDA_NoIndex_L005_R1.fastq -2 /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/Adult_filtered_6-25-2012-USDA_NoIndex_L005_R2.fastq -S adult.sam && samtools sort -@36 -o adult.bam adult.sam && rm adult.sam\n",
    "\n",
    "/gpfs/gpfs/scratch/xc278/p/HISTAT2/hisat2-2.1.0/hisat2 -p 36 -x Rdo -1 /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/Carcass-1_S3_L001_R1_001.fastq,/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/Carcass-2_S5_L001_R1_001.fastq,/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/Carcass-3_S9_L001_R1_001.fastq -2 /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/Carcass-1_S3_L001_R2_001.fastq,/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/Carcass-2_S5_L001_R2_001.fastq,/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/Carcass-3_S9_L001_R2_001.fastq -S carcass.sam && samtools sort -@36 -o carcass.bam carcass.sam && rm carcass.sam\n",
    "\n",
    "/gpfs/gpfs/scratch/xc278/p/HISTAT2/hisat2-2.1.0/hisat2 -p 36 -x Rdo -1 /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/Gut-1_S2_L001_R1_001.fastq,/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/Gut-2_S6_L001_R1_001.fastq,/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/Gut-3_S8_L001_R1_001.fastq -2 /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/Gut-1_S2_L001_R2_001.fastq,/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/Gut-2_S6_L001_R2_001.fastq,/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/Gut-3_S8_L001_R2_001.fastq -S gut.sam && samtools sort -@36 -o gut.bam gut.sam && rm gut.sam\n",
    "\n",
    "/gpfs/gpfs/scratch/xc278/p/HISTAT2/hisat2-2.1.0/hisat2 -p 36 -x Rdo -1 /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/Head-1_S1_L001_R1_001.fastq,/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/Head-2_S4_L001_R1_001.fastq,/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/Head-3_S7_L001_R1_001.fastq -2 /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/Head-1_S1_L001_R2_001.fastq,/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/Head-2_S4_L001_R2_001.fastq,/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/Head-3_S7_L001_R2_001.fastq\n",
    "-S head.sam && samtools sort -@36 -o head.bam head.sam && rm head.sam\n",
    "\n",
    "/gpfs/gpfs/scratch/xc278/p/HISTAT2/hisat2-2.1.0/hisat2 -p 36 -x Rdo -1 /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/SRR1130516_1.fastq -2 /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/SRR1130516_2.fastq -S srr.sam && samtools sort -@36 -o srr.bam srr.sam && rm srr.sam\n",
    "\n",
    "/gpfs/gpfs/scratch/xc278/p/HISTAT2/hisat2-2.1.0/hisat2 -p 36 -x Rdo -U /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/ResExp_Rep2.fastq,/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/ResExp_Rep3a.fastq,/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/ResExp_Rep3b.fastq,/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/ResExp_Rep3c.fastq,/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/ResExp_Rep4.fastq,/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/ResUxp_Rep2.fastq,/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/ResUxp_Rep3.fastq,/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/ResUxp_Rep4.fastq,/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/SuscExp_Rep2.fastq,/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/SuscExp_Rep3.fastq,/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/SuscExp_Rep4.fastq,/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/SuscUxp_Rep2.fastq,/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/SuscUxp_Rep3.fastq,/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/SuscUxp_Rep4.fastq -S unpair.sam && samtools sort -@36 -o unpair.bam unpair.sam && rm unpair.sam\n",
    "\n",
    "#stringtie\n",
    "/gpfs/gpfs/scratch/xc278/p/StringTie/stringtie-1.3.6.Linux_x86_64/stringtie -p 36 adult.bam -o adult.gtf\n",
    "/gpfs/gpfs/scratch/xc278/p/StringTie/stringtie-1.3.6.Linux_x86_64/stringtie -p 36 carcass.bam -o carcass.gtf\n",
    "/gpfs/gpfs/scratch/xc278/p/StringTie/stringtie-1.3.6.Linux_x86_64/stringtie -p 36 gut.bam -o gut.gtf\n",
    "/gpfs/gpfs/scratch/xc278/p/StringTie/stringtie-1.3.6.Linux_x86_64/stringtie -p 36 head.bam -o head.gtf\n",
    "/gpfs/gpfs/scratch/xc278/p/StringTie/stringtie-1.3.6.Linux_x86_64/stringtie -p 36 srr.bam -o srr.gtf\n",
    "/gpfs/gpfs/scratch/xc278/p/StringTie/stringtie-1.3.6.Linux_x86_64/stringtie -p 36 unpair.bam -o unpair.gtf\n",
    "\n",
    "#merge gff\n",
    "ls *.gtf >gtf_list\n",
    "/gpfs/gpfs/scratch/xc278/p/StringTie/stringtie-1.3.6.Linux_x86_64/stringtie --merge -o /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/StringTie/20190625RdoStringTie.gff gtf_list\n",
    "\n",
    "#get transcript sequence\n",
    "/gpfs/gpfs/scratch/xc278/p/cufflinks/cufflinks-2.2.1.Linux_x86_64/gffread -g /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/Genome/Rdo_Scaffolds.fa -w  /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/StringTie/20190625RdoStringTie.fa /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/StringTie/20190625RdoStringTie.gff\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MAKER"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "similar to https://reslp.github.io/blog/My-MAKER-Pipeline/\n",
    "\n",
    "Here, from BUSCO result, the quality of Genome is very high. Simplify the steps."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### split the genome to ~5kb and remove shorter than 200bp\n",
    "\n",
    "Note: for this genome, some bases are \"RYKMSWBDHV\", replace with N\n",
    "\n",
    "```python\n",
    "file_genome = '/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/Genome/Rdo_Scaffolds.fa'\n",
    "\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "import re\n",
    "\n",
    "n = 1\n",
    "l = 0\n",
    "outfolder = '/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/maker/scfs/'\n",
    "if not os.path.exists(outfolder):\n",
    "    os.makedirs(outfolder)\n",
    "outfile = open(os.path.join(outfolder,str(n)),'w')\n",
    "for s in SeqIO.parse(file_genome,'fasta'):\n",
    "    slen = len(s.seq)\n",
    "    if slen <=200:\n",
    "        continue\n",
    "    l += slen\n",
    "    seq = re.sub('[^ATCG]','N',str(s.seq))\n",
    "    outfile.write('>'+s.id+'\\n'+seq+'\\n')\n",
    "    if l >= 5000:\n",
    "        outfile.close()\n",
    "        n += 1\n",
    "        l = 0\n",
    "        outfile = open(os.path.join(outfolder,str(n)),'w')\n",
    "outfile.close()\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### combine EST sequences\n",
    "remove sequences with Non-ATCG bases, generate `/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/ref/ref.rna.faa`\n",
    "\n",
    "need to provide unique names, as there are some error reports from maker\n",
    "```python\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "\n",
    "files_nt = '''/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/StringTie/20190625RdoStringTie.fa\n",
    "/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/denovo/20190624Trinity.fasta\n",
    "/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/denovo/20190625transabyss.fasta\n",
    "'''.split()\n",
    "file_nt = '/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/ref/ref.rna.faa'\n",
    "fout = open(file_nt,'w')\n",
    "n = 0\n",
    "for f in files_nt:\n",
    "    for s in SeqIO.parse(f,'fasta'):\n",
    "        n += 1\n",
    "        seq = str(s.seq)\n",
    "        seq = seq.upper()\n",
    "        if seq.count('A')+seq.count('T')+seq.count('C')+seq.count('G')+seq.count('*')+seq.count('N') == len(seq):\n",
    "            fout.write('>EST'+str(n)+'\\n'+str(s.seq)+'\\n')\n",
    "        else:\n",
    "            print(s.id,'abnormal',s.seq)\n",
    "fout.close()\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### first round maker\n",
    "\n",
    "```bash\n",
    "source activate /gpfs/gpfs/scratch/xc278/maker2 #activate environment\n",
    "cd /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/maker/step1\n",
    "maker -CTL\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "modify the maker_opts.ctl file, changed lines are\n",
    "\n",
    "Because maker cannot handle gff file of StringTie, do not use it.\n",
    "\n",
    "```\n",
    "genome=/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/Genome/Rdo_Scaffolds.fa #genome sequence (fasta file or fasta embeded in GFF3 file)\n",
    "est=/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/ref/ref.rna.faa #set of ESTs or assembled mRNA-seq in fasta format\n",
    "protein=/gpfs/gpfs/staging/jx76-003/xc/20190613cooperationBeeGenomes/ref/uniprot_sprot.fasta  #protein sequence file in fasta format (i.e. from mutiple oransisms)\n",
    "est2genome=1 #infer gene predictions directly from ESTs, 1 = yes, 0 = no\n",
    "protein2genome=1 #infer predictions from protein homology, 1 = yes, 0 = no\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "#-----Genome (these are always required)\n",
    "genome=/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/Genome/Rdo_Scaffolds.fa #genome sequence (fasta file or fasta embeded in GFF3 file)\n",
    "organism_type=eukaryotic #eukaryotic or prokaryotic. Default is eukaryotic\n",
    "\n",
    "#-----Re-annotation Using MAKER Derived GFF3\n",
    "maker_gff= #MAKER derived GFF3 file\n",
    "est_pass=0 #use ESTs in maker_gff: 1 = yes, 0 = no\n",
    "altest_pass=0 #use alternate organism ESTs in maker_gff: 1 = yes, 0 = no\n",
    "protein_pass=0 #use protein alignments in maker_gff: 1 = yes, 0 = no\n",
    "rm_pass=0 #use repeats in maker_gff: 1 = yes, 0 = no\n",
    "model_pass=0 #use gene models in maker_gff: 1 = yes, 0 = no\n",
    "pred_pass=0 #use ab-initio predictions in maker_gff: 1 = yes, 0 = no\n",
    "other_pass=0 #passthrough anyything else in maker_gff: 1 = yes, 0 = no\n",
    "\n",
    "#-----EST Evidence (for best results provide a file for at least one)\n",
    "est=/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/ref/ref.rna.faa #set of ESTs or assembled mRNA-seq in fasta format\n",
    "altest= #EST/cDNA sequence file in fasta format from an alternate organism\n",
    "est_gff= #aligned ESTs or mRNA-seq from an external GFF3 file\n",
    "altest_gff= #aligned ESTs from a closly relate species in GFF3 format\n",
    "\n",
    "#-----Protein Homology Evidence (for best results provide a file for at least one)\n",
    "protein=/gpfs/gpfs/staging/jx76-003/xc/20190613cooperationBeeGenomes/ref/uniprot_sprot.fasta  #protein sequence file in fasta format (i.e. from mutiple oransisms)\n",
    "protein_gff=  #aligned protein homology evidence from an external GFF3 file\n",
    "\n",
    "#-----Repeat Masking (leave values blank to skip repeat masking)\n",
    "model_org=all #select a model organism for RepBase masking in RepeatMasker\n",
    "rmlib= #provide an organism specific repeat library in fasta format for RepeatMasker\n",
    "repeat_protein=/gpfs/gpfs/staging/jx76-003/xc/20190613cooperationBeeGenomes/ref/te_proteins.fasta #provide a fasta file of transposable element proteins for RepeatRunner\n",
    "rm_gff= #pre-identified repeat elements from an external GFF3 file\n",
    "prok_rm=0 #forces MAKER to repeatmask prokaryotes (no reason to change this), 1 = yes, 0 = no\n",
    "softmask=1 #use soft-masking rather than hard-masking in BLAST (i.e. seg and dust filtering)\n",
    "\n",
    "#-----Gene Prediction\n",
    "snaphmm= #SNAP HMM file\n",
    "gmhmm= #GeneMark HMM file\n",
    "augustus_species= #Augustus gene prediction species model\n",
    "fgenesh_par_file= #FGENESH parameter file\n",
    "pred_gff= #ab-initio predictions from an external GFF3 file\n",
    "model_gff= #annotated gene models from an external GFF3 file (annotation pass-through)\n",
    "est2genome=1 #infer gene predictions directly from ESTs, 1 = yes, 0 = no\n",
    "protein2genome=1 #infer predictions from protein homology, 1 = yes, 0 = no\n",
    "trna=0 #find tRNAs with tRNAscan, 1 = yes, 0 = no\n",
    "snoscan_rrna= #rRNA file to have Snoscan find snoRNAs\n",
    "unmask=0 #also run ab-initio prediction programs on unmasked sequence, 1 = yes, 0 = no\n",
    "\n",
    "#-----Other Annotation Feature Types (features MAKER doesn't recognize)\n",
    "other_gff= #extra features to pass-through to final MAKER generated GFF3 file\n",
    "\n",
    "#-----External Application Behavior Options\n",
    "alt_peptide=C #amino acid used to replace non-standard amino acids in BLAST databases\n",
    "cpus=1 #max number of cpus to use in BLAST and RepeatMasker (not for MPI, leave 1 when using MPI)\n",
    "\n",
    "#-----MAKER Behavior Options\n",
    "max_dna_len=1000000000 #length for dividing up contigs into chunks (increases/decreases memory usage)\n",
    "min_contig=1 #skip genome contigs below this length (under 10kb are often useless)\n",
    "\n",
    "pred_flank=200 #flank for extending evidence clusters sent to gene predictors\n",
    "pred_stats=0 #report AED and QI statistics for all predictions as well as models\n",
    "AED_threshold=1 #Maximum Annotation Edit Distance allowed (bound by 0 and 1)\n",
    "min_protein=0 #require at least this many amino acids in predicted proteins\n",
    "alt_splice=0 #Take extra steps to try and find alternative splicing, 1 = yes, 0 = no\n",
    "always_complete=0 #extra steps to force start and stop codons, 1 = yes, 0 = no\n",
    "map_forward=0 #map names and attributes forward from old GFF3 genes, 1 = yes, 0 = no\n",
    "keep_preds=0 #Concordance threshold to add unsupported gene prediction (bound by 0 and 1)\n",
    "\n",
    "split_hit=10000 #length for the splitting of hits (expected max intron size for evidence alignments)\n",
    "single_exon=1 #consider single exon EST evidence when generating annotations, 1 = yes, 0 = no\n",
    "single_length=600 #min length required for single exon ESTs if 'single_exon is enabled'\n",
    "correct_est_fusion=0 #limits use of ESTs in annotation to avoid fusion genes\n",
    "\n",
    "tries=2 #number of times to try a contig if there is a failure for some reason\n",
    "clean_try=0 #remove all data from previous run before retrying, 1 = yes, 0 = no\n",
    "clean_up=0 #removes theVoid directory with individual analysis files, 1 = yes, 0 = no\n",
    "TMP= #specify a directory other than the system default temporary directory for temporary files\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "generate scripts\n",
    "```python\n",
    "total_seg = 140\n",
    "folder = '/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/maker/step1/'\n",
    "commands = f'{folder}cmds.txt'\n",
    "fout = open(commands,'w')\n",
    "for n in range(1,total_seg):\n",
    "    cmd = f'source activate /gpfs/gpfs/scratch/xc278/maker2 && cd /home/scratch && mkdir {n} && cd {n} && cp {folder}/maker*.ctl ./ && maker -genome /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/maker/scfs/{n} -cpus 9 &&  gff3_merge -d *.maker.output/*master_datastore_index.log &&  mv *.gff {folder}/maker_split_run/ && rm -rf /home/scratch/{n} \\n'\n",
    "    fout.write(cmd)\n",
    "fout.close()\n",
    "```\n",
    "\n",
    "Note: change cpus to 9, to speed up a little bit.\n",
    "\n",
    "\n",
    "```bash\n",
    "srun -N 16 --partition long --qos long-award   -t 240:00:00 --pty bash #allocate 16 nodes\n",
    "python /home1/xc278/w/GitHub/xiaolongTools/multiThreadSlurm.py -t 9 -n e3c-012,e3c-013,e3c-014,e3c-015,e3c-016,e3c-017,e3c-018,e3c-019,e3c-020,e3c-021,e3c-022,e3c-023,e3c-024,e3c-025,e3c-026,e3c-027 -s 60 -i /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/maker/step1/cmds.txt \n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "some scaffolds are too large, too slow to run with single thread. tested, maker can be run with mpi with the code below. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "txt = '''#!/bin/bash\n",
    "\n",
    "#SBATCH --job-name={n}\n",
    "#SBATCH --partition=main\n",
    "#SBATCH --qos main-award \n",
    "#SBATCH --time=48:00:00\n",
    "#SBATCH --error=err.%x.%j.%N\n",
    "#SBATCH --output=out.%x.%j.%N\n",
    "#SBATCH --export=ALL\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=24\n",
    "####SBATCH --mem=256GB\n",
    "#SBATCH --mail-user=xiaolong.cao@rutgers.edu\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --requeue                    # Return job to the queue if preempted\n",
    "#SBATCH -A jx76-004 #which project acount to charge\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "source activate /gpfs/gpfs/scratch/xc278/maker2 && cd /home/scratch && mkdir {n} && cd {n} && cp /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/maker/step1//maker*.ctl ./ && mpiexec -n 36   maker -genome  /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/maker/scfs/{n}  &&  gff3_merge -d *.maker.output/*master_datastore_index.log &&  mv *.gff /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/maker/step1//maker_split_run/ && rm -rf /home/scratch/1 \n",
    "\n",
    "'''\n",
    "\n",
    "for i in '1 2 3 4 5 96 97 99 134'.split():\n",
    "    open(f\"{i}.sbatch\",'w',newline='\\n').write(txt.format(n=i))\n",
    "    \n",
    "\n",
    "```\n",
    "\n",
    "In the future, maybe use mpi maker directly. no need to split the genome, which will be much simpler."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "to be compatible with the following step, use python to combine the result\n",
    "\n",
    "```python\n",
    "import glob\n",
    "files = glob.glob('/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/maker/step1/maker_split_run/*.gff')\n",
    "outfile = '/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/maker/step1/step1.all.gff'\n",
    "\n",
    "ls_gff = []\n",
    "ls_fasta = []\n",
    "for f in files:\n",
    "    txt = open(f).read()\n",
    "    txt = txt.replace('##gff-version 3\\n','',1)\n",
    "    txt_gff, txt_fa = txt.split('##FASTA\\n')\n",
    "    ls_gff.append(txt_gff)\n",
    "    ls_fasta.append(txt_fa)\n",
    "\n",
    "fout = open(outfile,'w')\n",
    "fout.write('##gff-version 3\\n')\n",
    "fout.write(''.join(ls_gff))\n",
    "fout.write('##FASTA\\n')\n",
    "fout.write(''.join(ls_fasta))\n",
    "fout.close()\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### train SNAP"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "the scaffold name is like 'Scaffold_12;HRSCAF=78', has to change it to like `Scaffold_12_HRSCAF_78`, otherwise the program does not work normally\n",
    "\n",
    "```python\n",
    "file = 'step1.all.gff'\n",
    "txt = open(file).read()\n",
    "txt = txt.replace(';HRSCAF=','_HRSCAF_')\n",
    "txt = txt.replace('%3BHRSCAF%3D','_HRSCAF_')\n",
    "open(file,'w').write(txt)\n",
    "```\n",
    "\n",
    "```bash\n",
    "cd /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/maker/step1\n",
    "source activate /gpfs/gpfs/scratch/xc278/maker2 #activate environment\n",
    "\n",
    "maker2zff step1.all.gff\n",
    "fathom genome.ann genome.dna -validate > snap_validate_output.txt\n",
    "cat snap_validate_output.txt | grep \"error\" > error.log #5 error MODEL24893 MODEL2504 MODEL2505 MODEL918 MODEL1855\n",
    "grep -vwE \"(MODEL24893|MODEL2504|MODEL2505|MODEL918|MODEL1855)\" genome.ann > genome.ann2\n",
    "cat genome.ann2 |grep \"MODEL24893\"\n",
    "cat genome.ann |grep \"MODEL24893\"\n",
    "fathom genome.ann2 genome.dna -validate\n",
    "fathom genome.ann2 genome.dna -categorize 1000\n",
    "fathom uni.ann uni.dna -export 1000 -plus \n",
    "forge export.ann export.dna\n",
    "hmm-assembler.pl my_genome . > my_genome.hmm\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### step6 train augustus\n",
    "```bash\n",
    "cd /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/maker/step1\n",
    "perl /gpfs/gpfs/scratch/xc278/maker2/bin/zff2augustus_gbk.pl > augustus.gbk\n",
    "randomSplit.pl augustus.gbk 100\n",
    "new_species.pl --species=Rdo20190628\n",
    "etraining --species=Rdo20190628 augustus.gbk.train\n",
    "augustus --species=Rdo20190628 augustus.gbk.test | tee first_training.out\n",
    "optimize_augustus.pl --species=Rdo20190628 augustus.gbk.train --cpus=36 --kfold=36\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### train GeneMark\n",
    "```\n",
    "cd /lab02/Data_Raw/Xiaolong/temp\n",
    "export PERL5LIB=/home/xcao/p/anaconda3/lib/site_perl/5.26.2/\n",
    "perl /home/xcao/p/gm_et_linux_64/gmes_petap/gmes_petap.pl --cores 48 --ES --sequence Rdo_Scaffolds.fa\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### step9 maker final run\n",
    "\n",
    "```bash\n",
    "cd /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/maker/step9/\n",
    "cp /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/maker/step1/maker*.ctl ./\n",
    "#cp /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/Genome/Rdo_Scaffolds.fa ./\n",
    "#sed -i -e 's/=/_/g' -e 's/;/_/g' Rdo_Scaffolds.fa #change scaffold name, change '=' and ';' to '_'\n",
    "```\n",
    "change scaffold name, and change bases like \"Y\" in the sequence to \"N\"\n",
    "```python\n",
    "\n",
    "filein = '/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/Genome/Rdo_Scaffolds.fa'\n",
    "fileout = '/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/maker/step9/Rdo_Scaffolds.fa'\n",
    "\n",
    "from Bio import SeqIO\n",
    "import re\n",
    "\n",
    "fout = open(fileout,'w')\n",
    "for s in SeqIO.parse(filein,'fasta'):\n",
    "    slen = len(s.seq)\n",
    "    seq = re.sub('[^ATCG]','N',str(s.seq))\n",
    "    sid = re.sub('=|;','_', s.id)\n",
    "    fout.write('>'+sid+'\\n'+seq+'\\n')\n",
    "fout.close()\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "change the `maker_opts.ctl` to \n",
    "```\n",
    "#-----Genome (these are always required)\n",
    "genome=/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/maker/step9/Rdo_Scaffolds.fa #genome sequence (fasta file or fasta embeded in GFF3 file)\n",
    "organism_type=eukaryotic #eukaryotic or prokaryotic. Default is eukaryotic\n",
    "\n",
    "#-----Re-annotation Using MAKER Derived GFF3\n",
    "maker_gff=/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/maker/step1/step1.all.gff #MAKER derived GFF3 file\n",
    "est_pass=1 #use ESTs in maker_gff: 1 = yes, 0 = no\n",
    "altest_pass=1 #use alternate organism ESTs in maker_gff: 1 = yes, 0 = no\n",
    "protein_pass=1 #use protein alignments in maker_gff: 1 = yes, 0 = no\n",
    "rm_pass=1 #use repeats in maker_gff: 1 = yes, 0 = no\n",
    "model_pass=0 #use gene models in maker_gff: 1 = yes, 0 = no\n",
    "pred_pass=0 #use ab-initio predictions in maker_gff: 1 = yes, 0 = no\n",
    "other_pass=0 #passthrough anyything else in maker_gff: 1 = yes, 0 = no\n",
    "\n",
    "#-----EST Evidence (for best results provide a file for at least one)\n",
    "est= #set of ESTs or assembled mRNA-seq in fasta format\n",
    "altest= #EST/cDNA sequence file in fasta format from an alternate organism\n",
    "est_gff= #aligned ESTs or mRNA-seq from an external GFF3 file\n",
    "altest_gff= #aligned ESTs from a closly relate species in GFF3 format\n",
    "\n",
    "#-----Protein Homology Evidence (for best results provide a file for at least one)\n",
    "protein=  #protein sequence file in fasta format (i.e. from mutiple oransisms)\n",
    "protein_gff=  #aligned protein homology evidence from an external GFF3 file\n",
    "\n",
    "#-----Repeat Masking (leave values blank to skip repeat masking)\n",
    "model_org=all #select a model organism for RepBase masking in RepeatMasker\n",
    "rmlib= #provide an organism specific repeat library in fasta format for RepeatMasker\n",
    "repeat_protein= #provide a fasta file of transposable element proteins for RepeatRunner\n",
    "rm_gff= #pre-identified repeat elements from an external GFF3 file\n",
    "prok_rm=0 #forces MAKER to repeatmask prokaryotes (no reason to change this), 1 = yes, 0 = no\n",
    "softmask=1 #use soft-masking rather than hard-masking in BLAST (i.e. seg and dust filtering)\n",
    "\n",
    "#-----Gene Prediction\n",
    "snaphmm=/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/maker/step1/my_genome.hmm #SNAP HMM file\n",
    "gmhmm=/gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/maker/step8/gmhmm.mod #GeneMark HMM file\n",
    "augustus_species=Rdo20190628 #Augustus gene prediction species model\n",
    "fgenesh_par_file= #FGENESH parameter file\n",
    "pred_gff= #ab-initio predictions from an external GFF3 file\n",
    "model_gff= #annotated gene models from an external GFF3 file (annotation pass-through)\n",
    "est2genome=1 #infer gene predictions directly from ESTs, 1 = yes, 0 = no\n",
    "protein2genome=1 #infer predictions from protein homology, 1 = yes, 0 = no\n",
    "trna=0 #find tRNAs with tRNAscan, 1 = yes, 0 = no\n",
    "snoscan_rrna= #rRNA file to have Snoscan find snoRNAs\n",
    "unmask=0 #also run ab-initio prediction programs on unmasked sequence, 1 = yes, 0 = no\n",
    "\n",
    "#-----Other Annotation Feature Types (features MAKER doesn't recognize)\n",
    "other_gff= #extra features to pass-through to final MAKER generated GFF3 file\n",
    "\n",
    "#-----External Application Behavior Options\n",
    "alt_peptide=C #amino acid used to replace non-standard amino acids in BLAST databases\n",
    "cpus=1 #max number of cpus to use in BLAST and RepeatMasker (not for MPI, leave 1 when using MPI)\n",
    "\n",
    "#-----MAKER Behavior Options\n",
    "max_dna_len=100000 #length for dividing up contigs into chunks (increases/decreases memory usage)\n",
    "min_contig=1 #skip genome contigs below this length (under 10kb are often useless)\n",
    "\n",
    "pred_flank=200 #flank for extending evidence clusters sent to gene predictors\n",
    "pred_stats=0 #report AED and QI statistics for all predictions as well as models\n",
    "AED_threshold=1 #Maximum Annotation Edit Distance allowed (bound by 0 and 1)\n",
    "min_protein=0 #require at least this many amino acids in predicted proteins\n",
    "alt_splice=0 #Take extra steps to try and find alternative splicing, 1 = yes, 0 = no\n",
    "always_complete=0 #extra steps to force start and stop codons, 1 = yes, 0 = no\n",
    "map_forward=0 #map names and attributes forward from old GFF3 genes, 1 = yes, 0 = no\n",
    "keep_preds=0 #Concordance threshold to add unsupported gene prediction (bound by 0 and 1)\n",
    "\n",
    "split_hit=10000 #length for the splitting of hits (expected max intron size for evidence alignments)\n",
    "single_exon=1 #consider single exon EST evidence when generating annotations, 1 = yes, 0 = no\n",
    "single_length=600 #min length required for single exon ESTs if 'single_exon is enabled'\n",
    "correct_est_fusion=0 #limits use of ESTs in annotation to avoid fusion genes\n",
    "\n",
    "tries=2 #number of times to try a contig if there is a failure for some reason\n",
    "clean_try=0 #remove all data from previous run before retrying, 1 = yes, 0 = no\n",
    "clean_up=0 #removes theVoid directory with individual analysis files, 1 = yes, 0 = no\n",
    "TMP= #specify a directory other than the system default temporary directory for temporary files\n",
    "\n",
    "```\n",
    "\n",
    "run maker\n",
    "\n",
    "```bash\n",
    "mpiexec -n 36   maker\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### clean files\n",
    "\n",
    "```\n",
    "mv /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/maker/step9/Rdo_Scaffolds.* /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/20190622Rhyzopertha_dominica/\n",
    "cp /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/maker/step1/my_genome.hmm /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/20190622Rhyzopertha_dominica/Rdo.snap.hmm\n",
    "cp /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/maker/step8/gmhmm.mod /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/20190622Rhyzopertha_dominica/Rdo.genemark.hmm\n",
    "cp -r /gpfs/gpfs/scratch/xc278/maker2/config/species/Rdo20190628/ /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/20190622Rhyzopertha_dominica/\n",
    "pigz /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/20190622Rhyzopertha_dominica/Rdo_Scaffolds.all.gff\n",
    "\n",
    "rm -rf /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/RNA/\n",
    "rm -rf /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/maker/\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## BUSCO estimate quality"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### BUSCO for Genome"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```bash\n",
    "cd /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/BUSCO/\n",
    "source activate /gpfs/gpfs/scratch/xc278/maker2/\n",
    "python /home1/xc278/p/BUSCO/busco-master/scripts/run_BUSCO.py -c 36 -m genome -l /home1/xc278/p/BUSCO/insecta/insecta_odb9/ -i /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/Genome/Rdo_Scaffolds.fa -o 20190625RdoGenome\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "result\n",
    "```\n",
    "C:99.4%[S:98.8%,D:0.6%],F:0.2%,M:0.4%,n:1658\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### BUSCO for augustus protein\n",
    "\n",
    "```bash\n",
    "cd /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/BUSCO/\n",
    "source activate /gpfs/gpfs/scratch/xc278/maker2/\n",
    "python /home1/xc278/p/BUSCO/busco-master/scripts/run_BUSCO.py -c 36 -m prot -l /home1/xc278/p/BUSCO/insecta/insecta_odb9/ -i /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/Genome/blast2go_fasta_pred_prot.fasta -o 20190625RdoAugustusProt\n",
    "\n",
    "```\n",
    "\n",
    "result\n",
    "```\n",
    "C:71.9%[S:70.9%,D:1.0%],F:7.8%,M:20.3%,n:1658\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### BUSCO for StringTie, Trinity and TransAbyss"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```bash\n",
    "cd /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/BUSCO/\n",
    "source activate /gpfs/gpfs/scratch/xc278/maker2/\n",
    "python /home1/xc278/p/BUSCO/busco-master/scripts/run_BUSCO.py -c 36 -m tran -l /home1/xc278/p/BUSCO/insecta/insecta_odb9/ -i /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/StringTie/20190625RdoStringTie.fa -o 20190625Rdo.StringTie\n",
    "python /home1/xc278/p/BUSCO/busco-master/scripts/run_BUSCO.py -c 36 -m tran -l /home1/xc278/p/BUSCO/insecta/insecta_odb9/ -i /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/denovo/20190624Trinity.fasta -o 20190625Rdo.Trinity\n",
    "python /home1/xc278/p/BUSCO/busco-master/scripts/run_BUSCO.py -c 36 -m tran -l /home1/xc278/p/BUSCO/insecta/insecta_odb9/ -i /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/denovo/20190625transabyss.fasta -o 20190625Rdo.TransAbyss\n",
    "\n",
    "\n",
    "```\n",
    "\n",
    "result\n",
    "```\n",
    "StringTie: C:98.3%[S:57.0%,D:41.3%],F:0.8%,M:0.9%,n:1658\n",
    "Trinity: C:94.6%[S:58.3%,D:36.3%],F:4.6%,M:0.8%,n:1658\n",
    "TransAbyss: C:96.0%[S:94.9%,D:1.1%],F:3.7%,M:0.3%,n:1658\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### BUSCO for maker protein sets\n",
    "\n",
    "maker output 5 protein files. \n",
    "```\n",
    "Rdo_Scaffolds.all.maker.augustus_masked.proteins.fasta\n",
    "Rdo_Scaffolds.all.maker.genemark.proteins.fasta\n",
    "Rdo_Scaffolds.all.maker.non_overlapping_ab_initio.proteins.fasta\n",
    "Rdo_Scaffolds.all.maker.proteins.fasta\n",
    "Rdo_Scaffolds.all.maker.snap_masked.proteins.fasta\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```bash\n",
    "cd /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/BUSCO/\n",
    "source activate /gpfs/gpfs/scratch/xc278/maker2/\n",
    "python /home1/xc278/p/BUSCO/busco-master/scripts/run_BUSCO.py -c 36 -m prot -l /home1/xc278/p/BUSCO/insecta/insecta_odb9/ -i /gpfs/gpfs/staging/jx76-003/xc/20190622Rhyzopertha_dominica/20190622Rhyzopertha_dominica/Rdo_Scaffolds.all.maker.proteins.fasta -o 20190629RdoMakerProt\n",
    "\n",
    "```\n",
    "\n",
    "result\n",
    "```\n",
    "Rdo_Scaffolds.all.maker.proteins.fasta    C:92.1%[S:91.1%,D:1.0%],F:2.3%,M:5.6%,n:1658\n",
    "Rdo_Scaffolds.all.maker.augustus_masked.proteins.fasta        C:94.4%[S:93.3%,D:1.1%],F:3.6%,M:2.0%,n:1658\n",
    "Rdo_Scaffolds.all.maker.genemark.proteins.fasta        C:86.9%[S:85.8%,D:1.1%],F:7.8%,M:5.3%,n:1658\n",
    "Rdo_Scaffolds.all.maker.non_overlapping_ab_initio.proteins.fasta        C:2.8%[S:2.7%,D:0.1%],F:2.1%,M:95.1%,n:1658\n",
    "Rdo_Scaffolds.all.maker.snap_masked.proteins.fasta        C:72.2%[S:71.2%,D:1.0%],F:10.6%,M:17.2%,n:1658\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MCuNovo"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### translate with Transdecoder"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```bash\n",
    "cd /mnt/d/species/genomes/20190622Rhyzopertha_dominica/StringTie/\n",
    "~/p/transdecoder/TransDecoder-TransDecoder-v5.5.0/TransDecoder.LongOrfs -t /mnt/d/species/genomes/20190622Rhyzopertha_dominica/StringTie/20190625RdoStringTie.fa -m 60 -O /mnt/d/species/genomes/20190622Rhyzopertha_dominica/StringTie/transdecoder\n",
    "~/p/transdecoder/TransDecoder-TransDecoder-v5.5.0/TransDecoder.Predict -t /mnt/d/species/genomes/20190622Rhyzopertha_dominica/StringTie/20190625RdoStringTie.fa  -O /mnt/d/species/genomes/20190622Rhyzopertha_dominica/StringTie/transdecoder\n",
    "\n",
    "cd /mnt/d/species/genomes/20190622Rhyzopertha_dominica/denovo/\n",
    "~/p/transdecoder/TransDecoder-TransDecoder-v5.5.0/TransDecoder.LongOrfs -t /mnt/d/species/genomes/20190622Rhyzopertha_dominica/denovo/20190624Trinity.fasta -m 60 -O /mnt/d/species/genomes/20190622Rhyzopertha_dominica/denovo/20190624Trinity.transdecoder\n",
    "~/p/transdecoder/TransDecoder-TransDecoder-v5.5.0/TransDecoder.Predict -t /mnt/d/species/genomes/20190622Rhyzopertha_dominica/denovo/20190624Trinity.fasta  -O /mnt/d/species/genomes/20190622Rhyzopertha_dominica/denovo/20190624Trinity.transdecoder\n",
    "\n",
    "cd /mnt/d/species/genomes/20190622Rhyzopertha_dominica/denovo/\n",
    "~/p/transdecoder/TransDecoder-TransDecoder-v5.5.0/TransDecoder.LongOrfs -t /mnt/d/species/genomes/20190622Rhyzopertha_dominica/denovo/20190625transabyss.fasta -m 60 -O /mnt/d/species/genomes/20190622Rhyzopertha_dominica/denovo/20190625transabyss.transdecoder\n",
    "~/p/transdecoder/TransDecoder-TransDecoder-v5.5.0/TransDecoder.Predict -t /mnt/d/species/genomes/20190622Rhyzopertha_dominica/denovo/20190625transabyss.fasta  -O /mnt/d/species/genomes/20190622Rhyzopertha_dominica/denovo/20190625transabyss.transdecoder\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove duplicated protein sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```bash\n",
    "cd /mnt/d/species/genomes/20190622Rhyzopertha_dominica/StringTie/\n",
    "python /mnt/c/Users/ATPs/Documents/GitHub/MCuNovo/tools/fastaFileRemoveDup.py 20190625RdoStringTie.fa.transdecoder.pep\n",
    "\n",
    "cd /mnt/d/species/genomes/20190622Rhyzopertha_dominica/denovo/\n",
    "python /mnt/c/Users/ATPs/Documents/GitHub/MCuNovo/tools/fastaFileRemoveDup.py 20190624Trinity.fasta.transdecoder.pep\n",
    "\n",
    "cd /mnt/d/species/genomes/20190622Rhyzopertha_dominica/denovo/\n",
    "python /mnt/c/Users/ATPs/Documents/GitHub/MCuNovo/tools/fastaFileRemoveDup.py 20190625transabyss.fasta.transdecoder.pep\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### blast against uniprot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note: it seems blastp-fast is not so sensitive. use blastp if possible.\n",
    "```bash\n",
    "#download fasta file of arthropoda fasta from uniprot\n",
    "cd /mnt/d/species/uniprot/arthropoda\n",
    "wget ftp://ftp.ebi.ac.uk/pub/databases/fastafiles/uniprot/uniprotkb_arthropoda.gz\n",
    "gzip -d uniprotkb_arthropoda.gz\n",
    "\n",
    "#remove sequences that are labeled as fragment and with X\n",
    "python /mnt/c/Users/ATPs/Documents/GitHub/MCuNovo/tools/fastaFileClean.py -i uniprotkb_arthropoda\n",
    "\n",
    "#build database\n",
    "~/p/blast/ncbi-blast-2.9.0+/bin/segmasker -in uniprotkb_arthropodaclean -infmt fasta -parse_seqids -outfmt maskinfo_asn1_bin -out uniprotkb_arthropodaclean.asnb\n",
    "~/p/blast/ncbi-blast-2.9.0+/bin/makeblastdb -in uniprotkb_arthropodaclean -input_type fasta -dbtype prot -parse_seqids -mask_data uniprotkb_arthropodaclean.asnb -out uniprotkb_arthropodacleanDB -title 'uniprotkb_arthropodaclean20190714'\n",
    "\n",
    "#run blast\n",
    "~/p/blast/ncbi-blast-2.9.0+/bin/blastp -db /mnt/d/species/uniprot/arthropoda/uniprotkb_arthropodacleanDB -query /mnt/d/species/genomes/20190622Rhyzopertha_dominica/StringTie/20190625RdoStringTie.fa.transdecoder.pepUN -out /mnt/d/species/genomes/20190622Rhyzopertha_dominica/20190714MCuNovo/StringTie2U -outfmt 6 -num_threads 6 -max_target_seqs 20 -task blastp-fast  \n",
    "\n",
    "~/p/blast/ncbi-blast-2.9.0+/bin/blastp -db /mnt/d/species/uniprot/arthropoda/uniprotkb_arthropodacleanDB -query /mnt/d/species/genomes/20190622Rhyzopertha_dominica/denovo/20190624Trinity.fasta.transdecoder.pepUN -out /mnt/d/species/genomes/20190622Rhyzopertha_dominica/20190714MCuNovo/Trinity2U -outfmt 6 -num_threads 6 -max_target_seqs 20  -task blastp-fast  \n",
    "\n",
    "~/p/blast/ncbi-blast-2.9.0+/bin/blastp -db /mnt/d/species/uniprot/arthropoda/uniprotkb_arthropodacleanDB -query /mnt/d/species/genomes/20190622Rhyzopertha_dominica/denovo/20190625transabyss.fasta.transdecoder.pepUN -out /mnt/d/species/genomes/20190622Rhyzopertha_dominica/20190714MCuNovo/TransAbyss2U -outfmt 6 -num_threads 6 -max_target_seqs 20 -task blastp-fast  \n",
    "\n",
    "~/p/blast/ncbi-blast-2.9.0+/bin/blastp -db /mnt/d/species/uniprot/arthropoda/uniprotkb_arthropodacleanDB -query /mnt/d/species/genomes/20190622Rhyzopertha_dominica/20190622Rhyzopertha_dominica/Rdo_Scaffolds.all.maker.proteins.fasta -out /mnt/d/species/genomes/20190622Rhyzopertha_dominica/20190714MCuNovo/Maker2U -outfmt 6 -num_threads 6 -max_target_seqs 20 -task blastp-fast  \n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### parse blast6 to calculate match length\n",
    "\n",
    "```bash\n",
    "cd /mnt/d/species/genomes/20190622Rhyzopertha_dominica/20190714MCuNovo\n",
    "python /mnt/c/Users/ATPs/Documents/GitHub/MCuNovo/tools/parseBlast6.py -i /mnt/d/species/genomes/20190622Rhyzopertha_dominica/20190714MCuNovo/StringTie2U -o /mnt/d/species/genomes/20190622Rhyzopertha_dominica/20190714MCuNovo/StringTie2U.match -m 30 -q /mnt/d/species/genomes/20190622Rhyzopertha_dominica/StringTie/20190625RdoStringTie.fa.transdecoder.pepUN -s /mnt/d/species/uniprot/arthropoda/uniprotkb_arthropodaclean\n",
    "\n",
    "python /mnt/c/Users/ATPs/Documents/GitHub/MCuNovo/tools/parseBlast6.py -i /mnt/d/species/genomes/20190622Rhyzopertha_dominica/20190714MCuNovo/Trinity2U -o /mnt/d/species/genomes/20190622Rhyzopertha_dominica/20190714MCuNovo/Trinity2U.match -m 30 -q /mnt/d/species/genomes/20190622Rhyzopertha_dominica/denovo/20190624Trinity.fasta.transdecoder.pepUN -s /mnt/d/species/uniprot/arthropoda/uniprotkb_arthropodaclean\n",
    "\n",
    "python /mnt/c/Users/ATPs/Documents/GitHub/MCuNovo/tools/parseBlast6.py -i /mnt/d/species/genomes/20190622Rhyzopertha_dominica/20190714MCuNovo/TransAbyss2U -o /mnt/d/species/genomes/20190622Rhyzopertha_dominica/20190714MCuNovo/TransAbyss2U.match -m 30 -q /mnt/d/species/genomes/20190622Rhyzopertha_dominica/denovo/20190625transabyss.fasta.transdecoder.pepUN -s /mnt/d/species/uniprot/arthropoda/uniprotkb_arthropodaclean\n",
    "\n",
    "python /mnt/c/Users/ATPs/Documents/GitHub/MCuNovo/tools/parseBlast6.py -i /mnt/d/species/genomes/20190622Rhyzopertha_dominica/20190714MCuNovo/Maker2U -o /mnt/d/species/genomes/20190622Rhyzopertha_dominica/20190714MCuNovo/Maker2U.match -m 30 -q /mnt/d/species/genomes/20190622Rhyzopertha_dominica/20190622Rhyzopertha_dominica/Rdo_Scaffolds.all.maker.proteins.fasta -s /mnt/d/species/uniprot/arthropoda/uniprotkb_arthropodaclean\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### compare between four set of proteins\n",
    "```bash\n",
    "cd /mnt/d/species/genomes/20190622Rhyzopertha_dominica\n",
    "\n",
    "python /mnt/c/Users/ATPs/Documents/GitHub/MCuNovo/tools/SeqCompare.py -q ./20190622Rhyzopertha_dominica/Rdo_Scaffolds.all.maker.proteins.fasta -s ./StringTie/20190625RdoStringTie.fa.transdecoder.pepUN -o ./20190714MCuNovo/maker2StringTie -p ~/p/muscle/muscle\n",
    "\n",
    "python /mnt/c/Users/ATPs/Documents/GitHub/MCuNovo/tools/SeqCompare.py -q ./20190622Rhyzopertha_dominica/Rdo_Scaffolds.all.maker.proteins.fasta -s ./denovo/20190624Trinity.fasta.transdecoder.pepUN -o ./20190714MCuNovo/maker2Trinity -p ~/p/muscle/muscle\n",
    "\n",
    "python /mnt/c/Users/ATPs/Documents/GitHub/MCuNovo/tools/SeqCompare.py -q ./20190622Rhyzopertha_dominica/Rdo_Scaffolds.all.maker.proteins.fasta -s ./denovo/20190625transabyss.fasta.transdecoder.pepUN -o ./20190714MCuNovo/maker2TransAbyss -p ~/p/muscle/muscle\n",
    "\n",
    "python /mnt/c/Users/ATPs/Documents/GitHub/MCuNovo/tools/SeqCompare.py -q ./StringTie/20190625RdoStringTie.fa.transdecoder.pepUN -s ./denovo/20190624Trinity.fasta.transdecoder.pepUN -o ./20190714MCuNovo/StringTie2Trinity -p ~/p/muscle/muscle\n",
    "\n",
    "python /mnt/c/Users/ATPs/Documents/GitHub/MCuNovo/tools/SeqCompare.py -q ./StringTie/20190625RdoStringTie.fa.transdecoder.pepUN -s ./denovo/20190625transabyss.fasta.transdecoder.pepUN -o ./20190714MCuNovo/StringTie2TransAbyss -p ~/p/muscle/muscle\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### use blast to compare between four set of proteins\n",
    "\n",
    "use blast for reference in case there are some wired things. Not used for MCuNovo. To speed up, run on HPC\n",
    "```python\n",
    "import os\n",
    "\n",
    "os.chdir('/cache/home/xc278/w/GitHub/temp/')\n",
    "\n",
    "\n",
    "files = '''/cache/home/xc278/w/GitHub/temp/20190624Trinity.fasta.transdecoder.pepUN\n",
    "/cache/home/xc278/w/GitHub/temp/20190625RdoStringTie.fa.transdecoder.pepUN\n",
    "/cache/home/xc278/w/GitHub/temp/20190625transabyss.fasta.transdecoder.pepUN\n",
    "/cache/home/xc278/w/GitHub/temp/Rdo_Scaffolds.all.maker.proteins.fasta\n",
    "'''.split()\n",
    "\n",
    "keys = 'Trinity StringTie TransAbyss maker'.split()\n",
    "dc_files = dict(zip(keys, files))\n",
    "\n",
    "makeblastdb = '/cache/sw/packages/blast/2.6.0/ncbi-blast-2.6.0+/bin/makeblastdb'\n",
    "blastp = '/cache/sw/packages/blast/2.6.0/ncbi-blast-2.6.0+/bin/blastp'\n",
    "\n",
    "cmd_makedb = '''{makeblastdb} -in {protein_file} -input_type fasta -dbtype prot -parse_seqids -out {key} -title '{key}' &'''\n",
    "\n",
    "for k, protein_file in dc_files.items():\n",
    "    os.system(cmd_makedb.format(key=k, protein_file=protein_file, makeblastdb=makeblastdb))\n",
    "\n",
    "cmd_blast = '''cd /cache/home/xc278/w/GitHub/temp/ && {blastp} -db {subject} -query {protein_file} -out {query}2{subject} -outfmt 6 -num_threads 24 -task blastp-fast -max_target_seqs 20 &'''\n",
    "\n",
    "qs_pairs = []\n",
    "qs_pairs.append(['maker', 'StringTie'])\n",
    "qs_pairs.append(['maker', 'Trinity'])\n",
    "qs_pairs.append(['maker', 'TransAbyss'])\n",
    "qs_pairs.append(['StringTie', 'Trinity'])\n",
    "qs_pairs.append(['StringTie','TransAbyss'])\n",
    "for query, subject in qs_pairs:\n",
    "    print(cmd_blast.format(blastp=blastp, subject=subject, query=query, protein_file=dc_files[query]))\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### generate a table, with protein_id, transcript_id and gene_id\n",
    "\n",
    "```bash\n",
    "python /mnt/c/Users/ATPs/Documents/GitHub/MCuNovo/tools/Protein2Transcript2Gene.py -c /mnt/d/species/genomes/20190622Rhyzopertha_dominica//StringTie/20190625RdoStringTie.fa /mnt/d/species/genomes/20190622Rhyzopertha_dominica/StringTie/20190625RdoStringTie.fa.transdecoder.pepUN -o /mnt/d/species/genomes/20190622Rhyzopertha_dominica/20190714MCuNovo/protein2transcript2gene.txt -d /mnt/d/species/genomes/20190622Rhyzopertha_dominica/denovo/20190624Trinity.fasta.transdecoder.pepUN /mnt/d/species/genomes/20190622Rhyzopertha_dominica/denovo/20190625transabyss.fasta.transdecoder.pepUN -m /mnt/d/species/genomes/20190622Rhyzopertha_dominica/20190622Rhyzopertha_dominica/Rdo_Scaffolds.all.maker.proteins.fasta\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### generate a table of proteins with 'X'\n",
    "```bash\n",
    "python  /mnt/c/Users/ATPs/Documents/GitHub/MCuNovo/tools/proteinXFinder.py -i /mnt/d/species/genomes/20190622Rhyzopertha_dominica/StringTie/20190625RdoStringTie.fa.transdecoder.pepUN /mnt/d/species/genomes/20190622Rhyzopertha_dominica/20190622Rhyzopertha_dominica/Rdo_Scaffolds.all.maker.proteins.fasta -o /mnt/d/species/genomes/20190622Rhyzopertha_dominica/20190714MCuNovo/proteinXcount.txt\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### run selector.\n",
    "check the C:\\Users\\ATPs\\Documents\\GitHub\\MCuNovo\\tools\\SelectorBasedOnMatch.py file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### run BUSCO\n",
    "\n",
    "```\n",
    "cd /projectsp/f_jx76_1/xiaolong/temp\n",
    "python /cache/home/xc278/p/busco/scripts/run_BUSCO.py -c 24 -m prot -o 20190902  -l /cache/home/xc278/p/busco/insecta_odb9/ -i MCD.MCD.protein.txt\n",
    "```\n",
    "\n",
    "output\n",
    "\n",
    "```\n",
    "INFO    C:99.8%[S:90.0%,D:9.8%],F:0.1%,M:0.1%,n:1658\n",
    "INFO    1655 Complete BUSCOs (C)\n",
    "INFO    1492 Complete and single-copy BUSCOs (S)\n",
    "INFO    163 Complete and duplicated BUSCOs (D)\n",
    "INFO    2 Fragmented BUSCOs (F)\n",
    "INFO    1 Missing BUSCOs (M)\n",
    "\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "toc-autonumbering": true
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
